### ISSUE TITLE
Handling LoRA Scale in the Legacy LoRA Implementation.

### ISSUE THREAD

**Issue by dg845** [huggingface/diffusers#6756]
TL;DR Follow up to #6718. My suggestion is that for model blocks which use a legacy LoRA scale parameter `scale`, such as `ResnetBlock2D`, callers of these blocks like `CrossAttnDownBlock2D` should only supply the `scale` argument at inference time (e.g. when `not self.training`).

----

I am opening this issue to discuss the handling of the LoRA scale argument `scale` in the legacy LoRA implementation (e.g. when not using the `peft` backend). I'm not sure how relevant this will be given that the legacy LoRA implementation will eventually be deprecated.

To clarify my terminology, I will use the following notation: at training time the full LoRA forward pass for a weight matrix $`W_0`$ with non-LoRA forward pass $`h=W_0x`$ is 

```math
h = W_0x + \Delta W x = W_{0}x + \frac{\alpha}{r}BAx
```

where $`BA`$ is a low-rank adaptation of rank $`r`$ and $`\alpha`$ is the `lora_alpha` parameter. At inference time, the LoRA forward pass is

```math
h = W_0x + \eta\Delta W x = W_{0}x + \eta\frac{\alpha}{r}BAx
```

where we scale the LoRA update $`\Delta W`$ by an additional `lora_scale` parameter $`\eta`$. By this definition, `lora_scale` should only be supplied at inference time.

If we look at `src/diffusers/models/lora.py`, we can see that `LoRALinearLayer.forward` implements the training-time LoRA forward pass update $`\Delta W = \frac{\alpha}{r}BA`$ as above:

https://github.com/huggingface/diffusers/blob/7f58a76f485498518edb107666edd1b5a314f324/src/diffusers/models/lora.py#L220-L226

and `LoRACompatibleLinear.forward` implements the inference-time LoRA forward pass (using `LoRALinearLayer.forward`):

https://github.com/huggingface/diffusers/blob/7f58a76f485498518edb107666edd1b5a314f324/src/diffusers/models/lora.py#L433

where `scale` is the LoRA scale $`\eta`$.

However, if we look at the implementation of `CrossAttnDownBlock2D` we can see that the LoRA scale $`\eta`$ is used in the non-gradient checkpointing training forward pass, which I think is a bug:

https://github.com/huggingface/diffusers/blob/d4c7ab7bf1a00b8f416b3d20b77babac86f7fb44/src/diffusers/models/unets/unet_2d_blocks.py#L1197-L1198

Since $`\eta`$ is (correctly) not supplied to the gradient checkpointing forward pass:

https://github.com/huggingface/diffusers/blob/d4c7ab7bf1a00b8f416b3d20b77babac86f7fb44/src/diffusers/models/unets/unet_2d_blocks.py#L1183-L1188

this also has the effect of making the gradient checkpointing and non-gradient checkpointing forward passes at training time not equivalent.

So my suggestion is to only pass the `scale` argument to blocks which accept a LoRA scale argument like `ResnetDownBlock2D` at inference time:

```python
class CrossAttnDownBlock2D(nn.Module):
    ...
    def forward(...):
        ...
        lora_scale = (
            cross_attention_kwargs.get("scale", 1.0)
            if cross_attention_kwargs is not None and self.training
            else 1.0
        )
        ...
```

This should fix the training forward pass implementation when `lora_scale != 1.0` (since by supplying `lora_scale` in this case we're functionally using a different `lora_alpha` $`\alpha' = \eta\cdot\alpha`$ instead of the expected $`\alpha`$) and make it so that the gradient checkpointing and non-gradient checkpointing forward passes for training are equivalent.

**Comment by sayakpaul** [comment]
That makes sense! Thanks a lot for explaining this. I think for PEFT-induced LoRA training, the `alpha` is handled separately. More specifically, for non-PEFT LoRA training, we never supply any `alpha` during training. So, your solution looks good to me. 

@younesbelkada WDYT?

**Comment by sayakpaul** [comment]
@younesbelkada a gentle ping.

**Comment by rfan-debug** [comment]
Quick question, should we still use the rank as `lora_alpha` in DoRA training? It seems that the magnitude should absorb the lora_alpha term?

**Comment by sayakpaul** [comment]
@BenjaminBossan ^.

**Comment by sayakpaul** [comment]
@dg845 is this issue still relevant? We can completely deprecated the legacy LoRA implementation now.

**Comment by BenjaminBossan** [comment]
> Quick question, should we still use the rank as `lora_alpha` in DoRA training? It seems that the magnitude should absorb the `lora_alpha` term?

The scale is applied after DoRA normalization + magnitude. So it is not "normalized away" and would still affect the outcome.

**Comment by dg845** [comment]
@sayakpaul I think this issue involves only the legacy LoRA implementation so it is probably no longer relevant if the legacy implementation has been deprecated.

**Comment by sayakpaul** [comment]
Yeah.
